pilot/pkg/model/telemetry.go

```diff
@@ -0,0 +1,146 @@
+
+// Telemetry holds configuration for Telemetry API resources.
+type Telemetry struct {
+	Name      string         `json:"name"`
+	Namespace string         `json:"namespace"`
+	Spec      *tpb.Telemetry `json:"spec"`
+}
+
+// Telemetries organizes Telemetry configuration by namespace.
+type Telemetries struct {
+	// Maps from namespace to the Telemetry configus.
+	NamespaceToTelemetries map[string][]Telemetry `json:"namespace_to_telemetries"`
+
+	// The name of the root namespace.
+	RootNamespace string `json:"root_namespace"`
+}
+
+// GetTelemetries returns the Telemetry configurations for the given environment.
+func GetTelemetries(env *Environment) (*Telemetries, error) {
+	telemetries := &Telemetries{
+		NamespaceToTelemetries: map[string][]Telemetry{},
+		RootNamespace:          env.Mesh().GetRootNamespace(),
+	}
+
+	fromEnv, err := env.List(collections.IstioTelemetryV1Alpha1Telemetries.Resource().GroupVersionKind(), NamespaceAll)
+	if err != nil {
+		return nil, err
+	}
+	sortConfigByCreationTime(fromEnv)
+	for _, config := range fromEnv {
+		telemetry := Telemetry{
+			Name:      config.Name,
+			Namespace: config.Namespace,
+			Spec:      config.Spec.(*tpb.Telemetry),
+		}
+		telemetries.NamespaceToTelemetries[config.Namespace] =
+			append(telemetries.NamespaceToTelemetries[config.Namespace], telemetry)
+	}
+
+	return telemetries, nil
+}
+
+func (t *Telemetries) EffectiveTelemetry(namespace string, workload labels.Collection) *tpb.Telemetry {
+	if t == nil {
+		return nil
+	}
+
+	var effectiveSpec *tpb.Telemetry
+	if t.RootNamespace != "" {
+		effectiveSpec = t.namespaceWideTelemetry(t.RootNamespace)
+	}
+
+	if namespace != t.RootNamespace {
+		nsSpec := t.namespaceWideTelemetry(namespace)
+		if nsSpec != nil && effectiveSpec != nil {
+			effectiveSpec = shallowMerge(effectiveSpec, nsSpec)
+		}
+	}
+
+	for _, telemetry := range t.NamespaceToTelemetries[namespace] {
+		spec := telemetry.Spec
+		if len(spec.GetSelector().GetMatchLabels()) == 0 {
+			continue
+		}
+		selector := labels.Instance(spec.GetSelector().GetMatchLabels())
+		if workload.IsSupersetOf(selector) {
+			effectiveSpec = shallowMerge(effectiveSpec, spec)
+		}
+	}
+
+	return effectiveSpec
+}
+
+func (t *Telemetries) namespaceWideTelemetry(namespace string) *tpb.Telemetry {
+	for _, tel := range t.NamespaceToTelemetries[namespace] {
+		spec := tel.Spec
+		if len(spec.GetSelector().GetMatchLabels()) == 0 {
+			return spec
+		}
+	}
+	return nil
+}
+
+func shallowMerge(parent, child *tpb.Telemetry) *tpb.Telemetry {
+	return shallowMergeTracing(parent, child)
+}
+
+func shallowMergeTracing(parent, child *tpb.Telemetry) *tpb.Telemetry {
+	merged := parent.DeepCopy()
+
+	if len(child.GetTracing()) == 0 {
+		return merged
+	}
+
+	childCopy := child.DeepCopy()
+
+	if len(merged.GetTracing()) == 0 {
+		merged.Tracing = childCopy.Tracing
+		return merged
+	}
+
+	// only use the first Tracing for now (all that is suppported)
```

Comment:
Curious, what is the use case of the repeated tracing spec?

Reply:
It is possible that you would want to use it to set up provider-specific workload configuration in multiple places... but for `Tracing` that seems pretty dubious, given that we only support a single provider at the moment.

I think we left it as `repeated` to have it be similar to potential metrics and logging extensions. But we should reevaluate.

Reply:
I see. I think it is fine. It gives us max flexibility. Make non-repeated to repeated will be hard, if there is any chance it could be useful, I think we should just make it repeated.